File: petcare_mobile_con_otro/lib/api/notificaciones_service.dart
import 'dart:convert';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:petcare_mobile/models/notificacion.dart';

class NotificacionesService {
  final String? _apiUrl = dotenv.env['API_URL'];
  final _storage = const FlutterSecureStorage();

  Future<String?> _getToken() async {
    return await _storage.read(key: 'accessToken');
  }

  // GET /mis-notificaciones
  Future<List<Notificacion>> getMisNotificaciones() async {
    final token = await _getToken();
    final url = Uri.parse('$_apiUrl/notificaciones/mis-notificaciones');

    final response = await http.get(url, headers: {
      'Authorization': 'Bearer $token',
    });

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Notificacion.fromJson(json)).toList();
    } else {
      throw Exception('Error al cargar las notificaciones');
    }
  }

  // PATCH /:id/leida
  Future<void> markAsRead(String notificationId) async {
    final token = await _getToken();
    final url = Uri.parse('$_apiUrl/notificaciones/$notificationId/leida');

    await http.patch(url, headers: {
      'Authorization': 'Bearer $token',
    });
    // No nos preocupamos por la respuesta si es exitosa
  }

  // POST /marcar-todas-leidas
  Future<void> markAllAsRead() async {
    final token = await _getToken();
    final url = Uri.parse('$_apiUrl/notificaciones/marcar-todas-leidas');

    await http.post(url, headers: {
      'Authorization': 'Bearer $token',
    });
  }
}

File: petcare-mobile/src/app/core/notificaciones.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';

export interface Notificacion {
  _id: string;
  titulo: string;
  cuerpo: string;
  leida: boolean;
  createdAt: string;
  ruta?: string;
}

@Injectable({ providedIn: 'root' })
export class NotificacionesService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/notificaciones`;

  getMisNotificaciones(): Observable<Notificacion[]> {
    return this.http.get<Notificacion[]>(`${this.apiUrl}/mis-notificaciones`);
  }
  deleteNotificacion(id: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }
  markAsRead(id: string): Observable<any> {
    return this.http.patch(`${this.apiUrl}/${id}/leida`, {});
  }

  markAllAsRead(): Observable<any> {
    return this.http.post(`${this.apiUrl}/marcar-todas-leidas`, {});
  }
}